### PR TITLE
Make UEFI_BOOTLOADER work more generally

### PR DESCRIPTION
**Pull request by jsmeix**
Trying to make how UEFI_BOOTLOADER works
more generally usable as follow-up of
https://github.com/rear/rear/pull/1204

### CODE REVIEW

**Comment by jsmeix** [comment]
@gozora
hereby I did major changes in
rescue/default/850_save_sysfs_uefi_vars.sh
that (hopefully) make UEFI_BOOTLOADER
more generally usable.

I tried to keep all backward compatible
but I really need your careful review.

In particular I searched all ReaR scripts
for what variables are actually globally used
and I did not find that those variables that I
declared now as 'local' to be used elsewhere.

Furthermore I implemented a somewhat sophisticated
way how one same variable UEFI_BOOTLOADER
can be either specified as file to be used as UEFI bootloader
or as single string or array of filename globbing patterns
for 'find' to search for a file to be used as UEFI bootloader.
My implementation is based on the bash behaviour that
normal single string variables and arrays work compatible
in a very well reasonable way:
<pre>
# var="string"

# echo "$var"
string

# echo "${var[@]}"
string

# var=( 'this' 'that' )

# echo "$var"
this

# echo "${var[@]}"
this that

# var="something"

# echo "$var"
something

# echo "${var[@]}"
something that
</pre>

**Comment by gozora** [comment]
Hello @jsmeix,

I'm definitely going to try this out, hope you don't mind a small delay in this review because lately I have lot of (primary) work tasks even during weekend so it might take longer then usual ...

V.

**Comment by gdha** [comment]
@jsmeix serious change in code - testing will be easier then just looking at the code...I guess

**Comment by jsmeix** [comment]
@gdha 
I cannot test with reasonable effort all those various possible
scenarios (i.e. all those various Linux distributions) where
the right file for UEFI_BOOTLODER could be stored.

**Comment by jsmeix** [comment]
@gozora 
while sleeping over it I noticed that the current logic
is partially backwards - and I think we can get rid
of the BugError - I will do another commit...

**Comment by gozora** [comment]
@jsmeix thanks for notification!
I was planing to take a look on it today afternoon/evening ...

V.

**Comment by jsmeix** [comment]
@gozora 
I think now you could have a look.

Unfortunately it became somewhat nested conditions but
currently I have no good idea how to make it less nested
(without using things like artificial helper variables).

**Comment by jsmeix** [comment]
Whoops!
I had the wrong assumption that any UEFI related code
belongs to @gozora until now when I used
<pre>
git log -p --follow usr/share/rear/rescue/default/850_save_sysfs_uefi_vars.sh
</pre>
that shows me that @gdha had initially made it
so that hereby I humbly also request a review
from @gdha

**Comment by jsmeix** [comment]
Caution:
In general
<pre>
is_true $VAR
</pre>
is not the logical opposite (i.e. the logical negation) of
<pre>
is_false $VAR
</pre>
see my explanation in usr/share/rear/lib/global-functions.sh
<pre>
# two explicit functions to be able to test explicitly for true and false (see issue #625)
# because "tertium non datur" (cf. https://en.wikipedia.org/wiki/Law_of_excluded_middle)
# does not hold for variables because variables could be unset or have empty value
...
    # only if there is explicitly a 'true' value then is_true returns true
    # so that an unset variable or an empty value is not true
...
    # only if there is explicitly a 'false' value then is_false returns true
    # so that an unset variable or an empty value is not false
    # caution: for unset or empty variables is_false is false
</pre>

File: usr/share/rear/rescue/default/850_save_sysfs_uefi_vars.sh [hunk 1]
```diff
@@ -1,84 +1,97 @@
...
-fi
+# A simplified uefivars replacement
 
-# Perhaps we defined UEFI_BOOTLOADER in /etc/rear/local.conf - check this now
-if [[ ! -z "${UEFI_BOOTLOADER}" ]]; then
-    # right, variable is not empty, but is it a file?
-    [[ ! -f ${UEFI_BOOTLOADER} ]] && Error "Cannot find a proper UEFI_BOOTLOADER ($UEFI_BOOTLOADER). 
-Please define it in $CONFIG_DIR/local.conf (e.g. UEFI_BOOTLOADER=/boot/efi/EFI/fedora/grubx64.efi)"
+# USING_UEFI_BOOTLOADER empty or 0 means NO UEFI
+is_true $USING_UEFI_BOOTLOADER || return
```
Comment (jsmeix) [inline usr/share/rear/rescue/default/850_save_sysfs_uefi_vars.sh]: See my reply directly in the pull request conversation
https://github.com/rear/rear/pull/1212#issuecomment-283698753

**Comment by jsmeix** [comment]
Basically directly after leaving the office yesterday
I remembered what I already did to avoid
code with deeply nested 'if...else' conditions:
I use an artificial 'for' clause that is run only once
so that I can at any point 'continue' with the code
after that 'for' clause, cf.
https://github.com/rear/rear/blob/master/usr/share/rear/restore/default/990_move_away_restored_files.sh

Since my above
https://github.com/rear/rear/pull/1212/commits/9a31a5fa9a75a72bfb01eedabadc104400982667
I even like to look at my code ;-)

@gozora 
many thanks for your tests!

I think I will merge it soon - unless there are objections.

@gdha 
of course when there are bugs, I will fix them.

**Pull request merged**